hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/OMDBInsightSearchEndpoint.java

```diff
@@ -0,0 +1,457 @@
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.io.IOException;
+import java.util.Map;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Set;
+
+import static org.apache.hadoop.ozone.OzoneConsts.OM_KEY_PREFIX;
+import static org.apache.hadoop.ozone.recon.ReconConstants.*;
+import static org.apache.hadoop.ozone.recon.api.handlers.BucketHandler.getBucketHandler;
+import static org.apache.hadoop.ozone.recon.api.handlers.EntityHandler.normalizePath;
+import static org.apache.hadoop.ozone.recon.api.handlers.EntityHandler.parseRequestPath;
+
+/**
+ * REST endpoint for search implementation in OM DB Insight.
+ */
+@Path("/keys")
+@Produces(MediaType.APPLICATION_JSON)
+@AdminOnly
+public class OMDBInsightSearchEndpoint {
+
+  private OzoneStorageContainerManager reconSCM;
+  private final ReconOMMetadataManager omMetadataManager;
+  private static final Logger LOG =
+      LoggerFactory.getLogger(OMDBInsightSearchEndpoint.class);
+  private ReconNamespaceSummaryManagerImpl reconNamespaceSummaryManager;
+
+
+  @Inject
+  public OMDBInsightSearchEndpoint(OzoneStorageContainerManager reconSCM,
+                                   ReconOMMetadataManager omMetadataManager,
+                                   ReconNamespaceSummaryManagerImpl reconNamespaceSummaryManager) {
+    this.reconSCM = reconSCM;
+    this.omMetadataManager = omMetadataManager;
+    this.reconNamespaceSummaryManager = reconNamespaceSummaryManager;
+  }
+
+
+  /**
+   * Performs a search for open keys in the Ozone Manager (OM) database using a specified search prefix.
+   * This endpoint searches across both File System Optimized (FSO) and Object Store (non-FSO) layouts,
+   * compiling a list of keys that match the given prefix along with their data sizes.
+   * <p>
+   * The search prefix may range from the root level ('/') to any specific directory
+   * or key level (e.g., '/volA/' for everything under 'volA'). The search operation matches
+   * the prefix against the start of keys' names within the OM DB.
+   * <p>
+   * Example Usage:
+   * 1. A startPrefix of "/" will return all keys in the database.
+   * 2. A startPrefix of "/volA/" retrieves every key under volume 'volA'.
+   * 3. Specifying "/volA/bucketA/dir1" focuses the search within 'dir1' inside 'bucketA' of 'volA'.
+   *
+   * @param startPrefix The prefix for searching keys, starting from the root ('/') or any specific path.
+   * @param limit       Limits the number of returned keys.
+   * @return A KeyInsightInfoResponse, containing matching keys and their data sizes.
+   * @throws IOException On failure to access the OM database or process the operation.
+   */
+  @GET
+  @Path("/open/search")
+  public Response searchOpenKeys(
+      @DefaultValue(DEFAULT_START_PREFIX) @QueryParam("startPrefix")
+      String startPrefix,
+      @DefaultValue(RECON_OPEN_KEY_DEFAULT_SEARCH_LIMIT) @QueryParam("limit")
```

Comment:
The current functionality of the limit is such that let's say the endpoint searches the entire filesystem for a specific key. After aggregating the search results, it then applies a limit to the number of records returned. The purpose of this limit is to provide a safeguard, especially in cases where the user might provide a broad path, such as "`/`", which could potentially return all open files in the system. For instance, if there are 1 million open files, returning all these results could significantly slow down the system and disrupt the user interface. The default limit is set at 1,000 records. However, users have the option to adjust this limit by specifying a higher value when calling the endpoint.

Reply:
@ArafatKhan2198 whenever we provide limit and startPrefix for any API, we should always provide user to get another page. Pls check , in this API, if user limits the results to 100 and still 100 more results to show in that given startPrefix, there is no way for user to get another page, so I believe providing lastKey is also needed here in response.

Reply:
@ArafatKhan2198 @devmadhuu what was the conclusion here? It makes sense what Devesh suggested.

Reply:
@ArafatKhan2198 As discussed, pls update the code to provide search at bucket level, not root or volume level, else providing last key for both fso and non-fso becomes too complex to traverse and confusing.
Since this is key search and keys are starting from bucket level, so it makes sense to allow user to search min from bucket level and then bucket can be either fso or non-fso which will make it easy for you to provide lastKey in response for pagination.

Reply:
Thank you for the review, @dombizita and @devmadhuu. As discussed, we'll restrict searches to start from the bucket level and proceed deeper. We won't initiate searches from the volume level or root level. Additionally, we'll include the last key in the response to facilitate pagination.